src/test/shell/bazel/bazel_rules_test.sh

```diff
@@ -743,4 +743,153 @@ EOF
     bazel run //pkg:my_executable >$TEST_log 2>&1 || fail "Binary should have exit code 0"
 }
 
+function test_starlark_rule_with_run_under_env() {
+  mkdir pkg
+  cat >pkg/BUILD <<'EOF'
+load(":rules.bzl", "my_executable", "my_test", "my_wrapper")
+my_executable(
+  name = "my_executable",
+)
+my_test(
+  name = "my_test",
+)
+my_wrapper(
+  name = "my_wrapper",
+)
+EOF
+
+  # On Windows this file needs to be acceptable by CreateProcessW(), rather
+  # than a Bourne script.
+  if "$is_windows"; then
+    local run_under_script="pkg/run_under.bat"
+    echo 'exports_files(["run_under.bat"])' >> pkg/BUILD
+    cat >pkg/run_under.bat <<'EOF'
+@ECHO OFF
+if not defined RUN_UNDER_VALUE (
+  echo --run_under script was unexpectedly executed
+)
+exit /B 101
+EOF
+    cat >pkg/rules.bzl <<'EOF'
+_SCRIPT_EXT = ".bat"
+_SCRIPT_CONTENT = """
+if not defined RUN_UNDER_VALUE (
+  echo run_under_env value is not set
+  exit /B 1
+)
+start /b /wait bash.exe -c "%RUN_UNDER_VALUE%"
+if %ERRORLEVEL% NEQ 101 (
+  echo Failed to execute run_under_env value
+  exit /B 1
+)
+exit /B 0
+"""
+EOF
+  else
+  local run_under_script="pkg/run_under.sh"
+  echo 'exports_files(["run_under.sh"])' >> pkg/BUILD
+    cat >pkg/run_under.sh <<'EOF'
+#!/bin/bash
+if [[ -z "${RUN_UNDER_VALUE}" ]]; then
+  echo --run_under script was unexpectedly executed
+fi
+exit 101
+EOF
+    chmod +x pkg/run_under.sh
+    cat >pkg/rules.bzl <<'EOF'
+_SCRIPT_EXT = ".sh"
+_SCRIPT_CONTENT = """#!/bin/bash
+if [[ -z "${RUN_UNDER_VALUE:-}" ]]; then
+  echo "run_under_env value is not set"
+  exit 1
+fi
+${RUN_UNDER_VALUE}
+if [[ $? != 101 ]]; then
+  echo "Failed to execute run_under_env value"
+  exit 1
+fi
+exit 0
+"""
+EOF
+  fi
+
+  cat >>pkg/rules.bzl <<'EOF'
+def _my_executable_impl(ctx):
+    executable_sh = ctx.actions.declare_file(ctx.attr.name + _SCRIPT_EXT)
+    ctx.actions.write(
+        output = executable_sh,
+        content = _SCRIPT_CONTENT,
+        is_executable = True,
+    )
+    return [
+        DefaultInfo(
+            executable = executable_sh,
+        ),
+        RunEnvironmentInfo(
+            run_under_env = "RUN_UNDER_VALUE",
+        ),
+    ]
+
+my_executable = rule(
+    implementation = _my_executable_impl,
+    attrs = {},
+    executable = True,
+)
+
+my_test = rule(
+    implementation = _my_executable_impl,
+    attrs = {},
+    test = True,
+)
+
+def _my_wrapper_impl(ctx):
+    executable_sh = ctx.actions.declare_file(ctx.attr.name + _SCRIPT_EXT)
+    ctx.actions.expand_template(
```

Comment:
With no substitutions, you should be able to use `ctx.actions.write` instead.

Reply:
I wanted to use `pkg/run_under.(sh|bat)` directly instead of needing to render the same content into a starlark constant to be written in this rule. This is one of the few actions that let me copy files.

Reply:
True, I missed that point!

Reply:
I replaced it with `ct.actions.symlink` which I was originally nervous about using because of windows but that seems to work 😄

Reply:
It will do a copy if symlinks aren't enabled. Keep calm and symlink on 🙂